### ISSUE TITLE
Add the ability to remotely sign repository

### ISSUE THREAD

**Issue by KyleFromKitware** [aptly-dev/aptly#839]
Aptly should have the ability to sign its repositories by SSH-ing to another machine and running GPG from there. That way, the GPG key and the repository management can live on separate servers, reducing the risk of the key being compromised (the signing server can have all network access disabled except for allowing the GPG command to be run by Aptly.)

**Comment by KyleFromKitware** [comment]
More generally, it would be nice to be able to specify the command to use instead of `gpg`. Then the user could write a wrapper that runs SSH when Aptly calls it.

**Comment by jdelic** [comment]
I think from aptly's side allowing the user to specify the exact command to run would suffice. That would also have prevented the whole `gpg` vs `gpg2` issue.

It's unrelated to aptly, but on your side, I would strongly recommend to instead use [Vault](https://vaultproject.io) and [vault-gpg-plugin](https://github.com/LeSuisse/vault-gpg-plugin) to get proper auditing and authentication, instead of rolling your own with SSH.

**Comment by KyleFromKitware** [comment]
> It's unrelated to aptly, but on your side, I would strongly recommend to instead use Vault and vault-gpg-plugin to get proper auditing and authentication, instead of rolling your own with SSH.

Yes, I agree that a comprehensive solution would be better than a hacked SSH script... haven't looked at Vault, I'll have to check it out.

**Comment by smira** [comment]
yes, interesting idea... I wonder what would be the best solution here, command which works as a filter with some arguments to specify key ID and other options

**Comment by KyleFromKitware** [comment]
I think we also need a way to distinguish between the *signing* provider and the *verification* provider. Verification is trivial and requires no handling of sensitive keys, and so most (all?) users can just use one of the existing `gpg` or `internal` providers for this purpose, even if some custom provider is used for signing.

**Comment by KyleFromKitware** [comment]
Something like:

```json
{
  "gpgSigningProvider": "custom",
  "gpgSigningCommand": "/path/to/signing/command",
  "gpgVerificationProvider": "gpg"
}
```

The trick with the custom signing command is coming up with a protocol that is both

1. extensible for future improvements
2. easy to process by a shell script or similar

As a naive approach, I'm imagining a command that takes the key ID as a single argument, accepts the to-be-signed file on standard input, and spits out the signed file on standard input, but this doesn't leave much room for future improvement :( on the other hand I don't want to make my shell script process *every single argument* that GPG recognizes.
